src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDirectoryValue.java

```diff
@@ -14,176 +14,72 @@
-
-  /** Represents a successful repository lookup. */
-  public static final class SuccessfulRepositoryDirectoryValue extends RepositoryDirectoryValue {
-    private final Path path;
-    private final boolean fetchingDelayed;
-    private final boolean excludeFromVendoring;
-    @Nullable private final byte[] digest;
-    @Nullable private final DirectoryListingValue sourceDir;
-    private final ImmutableMap<SkyKey, SkyValue> fileValues;
-
-    private SuccessfulRepositoryDirectoryValue(
-        Path path,
-        boolean fetchingDelayed,
-        @Nullable DirectoryListingValue sourceDir,
-        byte[] digest,
-        ImmutableMap<SkyKey, SkyValue> fileValues,
-        boolean excludeFromVendoring) {
-      this.path = path;
-      this.fetchingDelayed = fetchingDelayed;
-      this.sourceDir = sourceDir;
-      this.digest = digest;
-      this.fileValues = fileValues;
-      this.excludeFromVendoring = excludeFromVendoring;
-    }
-
+  record Success(Path path, boolean isFetchingDelayed, boolean excludeFromVendoring)
+      implements RepositoryDirectoryValue {
 
     @Override
     public boolean repositoryExists() {
       return true;
     }
 
-    @Override
-    public String getErrorMsg() {
-      throw new IllegalStateException();
-    }
-
-    @Override
     public Path getPath() {
       return path;
     }
-
-    @Override
-    public boolean isFetchingDelayed() {
-      return fetchingDelayed;
-    }
-
-    @Override
-    public boolean excludeFromVendoring() {
-      return excludeFromVendoring;
-    }
-
-    @Override
-    public boolean equals(Object other) {
-      if (this == other) {
-        return true;
-      }
-
-      if (other instanceof SuccessfulRepositoryDirectoryValue otherValue) {
-        return Objects.equal(path, otherValue.path)
-            && Objects.equal(sourceDir, otherValue.sourceDir)
-            && Arrays.equals(digest, otherValue.digest)
-            && Objects.equal(fileValues, otherValue.fileValues)
-            && Objects.equal(excludeFromVendoring, otherValue.excludeFromVendoring);
-      }
-      return false;
-    }
-
-    @Override
-    public int hashCode() {
-      return Objects.hashCode(path, sourceDir, Arrays.hashCode(digest), fileValues);
-    }
-
-    @Override
-    public String toString() {
-      return path.getPathString();
-    }
   }
 
-  /** Represents an unsuccessful repository lookup. */
-  public static final class NoRepositoryDirectoryValue extends RepositoryDirectoryValue {
-    private final String errorMsg;
-
-    public NoRepositoryDirectoryValue(String errorMsg) {
-      this.errorMsg = errorMsg;
-    }
+  /**
+   * Represents an unsuccessful repository lookup.
+   *
+   * @param errorMsg For an unsuccessful repository lookup, gets a detailed error message that is
+   *     suitable for reporting to a user.
+   */
+  record Failure(String errorMsg) implements RepositoryDirectoryValue {
 
     @Override
     public boolean repositoryExists() {
```

Comment:
Could this be replaced by `instanceof` checks and be deleted?

Reply:
done.